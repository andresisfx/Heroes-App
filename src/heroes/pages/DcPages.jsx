import React from 'react'
import { HeroList } from '../components/HeroList'

export const DcPages = () => {
  return (
    <>
      <h1>DcPages</h1> 
      <HeroList publisher="DC Comics" />
    </>
  )
}
