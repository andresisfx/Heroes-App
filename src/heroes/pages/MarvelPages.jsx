import React from 'react'
import { HeroList } from '../components'
export const MarvelPages = () => {
  return (
    <>
      <h1>DcPages</h1> 
           <HeroList publisher="Marvel Comics" />
    </>
  )
}
