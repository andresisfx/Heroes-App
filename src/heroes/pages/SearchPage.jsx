import React, { useMemo } from 'react'
import queryString from 'query-string'
import { HeroCard } from '../components'
import { useForm } from '../../hooks/useForm'
import { useNavigate,useLocation } from 'react-router-dom'
import { getHeroesByName } from '../helpers'

export const SearchPage = () => {
    
   const naviagte= useNavigate();
   const location=useLocation();

   const {q=''}=queryString.parse(location.search);

    const {searchText, onInputChange, onResetForm}=useForm(
        {
            searchText:''
        }
    );

    const onSearchSubmit=(event)=>{
        event.preventDefault();
        if(searchText.trim().length<=1) return;
        naviagte(`?q=${searchText.toLowerCase().trim()}`);
    }

    const heroes=getHeroesByName(q);

    const showSearch=(q.length===0);
    const showError=(q.length>0 && heroes.length===0);
    console.log("error",showError);
    console.log("search",showSearch);

  return (
    <>
      <h1>Seacrh</h1>
      <hr />
      <div className='row'>
            <div className='col-5'>

                <form action="">
                <input
                    type="text"
                    className='form-control'
                    placeholder='Find your hero'
                    aria-label="Find your hero"
                    name='searchText'
                    autoComplete='off'
                    value={searchText}
                    onChange={onInputChange}
                />
                <button
                    type='submit'
                    className='btn btn-outline-primary mt-1'
                    onClick={onSearchSubmit}

                >Search</button>
                </form>

            </div>
            <div className='col-7'>
                <h4>Results</h4>
                <hr />
                <div className='alert alert-primary animate__animated animate__fadeIn' style={{display: showSearch ? '' : 'none'}}>
                    Search Hero
                </div>

                <div className='alert alert-danger animate__animated animate__fadeIn'style={{display: showError ? '' : 'none' }}>
                    No hero with <b>{q}</b>
                </div>
                

                {
                  heroes?.map(hero=>(
                  <HeroCard {...hero} key={hero.id}
                  />))
                }
                
            </div>
      </div>
    </>
  )
}
