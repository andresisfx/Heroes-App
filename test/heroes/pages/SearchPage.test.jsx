import { MemoryRouter } from "react-router-dom";
import { render ,screen,fireEvent} from "@testing-library/react";
import { SearchPage } from "../../../src/heroes/pages/SearchPage";


const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
        ...jest.requireActual('react-router-dom'),
        useNavigate: () => mockNavigate,
    }));

describe('Pruebas en SearchPage', () => {

      beforeEach(() => jest.clearAllMocks());

    test('debe de mostrar el componente', () => {
        
        
     const {container}=   render(
              <MemoryRouter>
                    <SearchPage/>
              </MemoryRouter>
        );
        expect(container).toMatchSnapshot()
        
    })

    test('debe de mostrar a Batman y el input con el valor de q', () => {
        
        render(
              <MemoryRouter initialEntries={['/search?q=batman']}>
                    <SearchPage/>
              </MemoryRouter>
        );
        const inputValue=screen.getByRole('textbox')
        expect( inputValue.value).toBe('batman');

        const img=screen.getByRole('img');
        expect(img.src).toContain('batman');

        const alert=screen.getByLabelText('alert-danger');
        expect(alert.style.display).toBe('none');
    })

    test('debe de mostrar un error si el heroe no existe', () => {
        
        render(
              <MemoryRouter initialEntries={['/search?q=batman123']}>
                    <SearchPage/>
              </MemoryRouter>
        );
        const alert=screen.getByLabelText('alert-danger');
        expect(alert.style.display).toBe('');
    })

    test('debe de llamar el navigate', () => {
        
       
        render(
              <MemoryRouter initialEntries={['/search']}>
                    <SearchPage navigate={mockNavigate}/>
              </MemoryRouter>
        );
        const inputValue=screen.getByRole('textbox')
        fireEvent.change(inputValue,{target:{value:'superman'}})
        const form=screen.getByRole('form');
        fireEvent.submit(form);
        expect(mockNavigate).toHaveBeenCalledWith('?q=superman');
        
    })
})