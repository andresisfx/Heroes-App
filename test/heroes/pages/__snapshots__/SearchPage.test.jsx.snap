// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en SearchPage debe de mostrar el componente 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <form
        aria-label="form"
      >
        <input
          aria-label="Find your hero"
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-1"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        class="alert alert-primary animate__animated animate__fadeIn"
      >
        Search Hero
      </div>
      <div
        aria-label="alert-danger"
        class="alert alert-danger animate__animated animate__fadeIn"
        style="display: none;"
      >
        No hero with 
        <b />
      </div>
    </div>
  </div>
</div>
`;
